+str #19390 Add FlowMonitor

talp-traiana opened the pull request:
Add flow that materializes into object that provides the status of the flow (last received message, failures, etc.)

Still no tests/docs.

talpr committed:
+str #19390 Add 'monitor' flow combinator and stage

johanandren commented:
Refs #19390

johanandren commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -69,6 +66,9 @@ object Flow {
     combine: function.Function2[M1, M2, M]): Flow[I, O, M] =
     new Flow(scaladsl.Flow.fromSinkAndSourceMat(sink, source)(combinerToScala(combine)))
+
+  def monitorFlow[T](): Flow[T, T, FlowMonitor[T]] =
```
I think it should have the same name in all three places. `monitor` is good in my opinion.

johanandren commented:
Needs test coverage, doc on the public methods and mention in stages-overview for Scala and Java in the docs.

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala:
```diff
@@ -342,4 +338,6 @@ object Sink {
   def queue[T](): Sink[T, SinkQueue[T]] =
     Sink.fromGraph(new QueueSink())
 
+  def monitored[T, M](inner: Sink[T, M]): Sink[T, (M, FlowMonitor[T])] =
```
Actually, I don`t think it should be on `Sink` as `flow.to(sink)` becomes a sink.

talp-traiana replied:
So what should I do with it? Leave it as is? Move it someplace? Remove it?

johanandren replied:
I think you should remove it and keep it on `Flow` and instances of flow.

talp-traiana commented on akka-actor/src/main/scala/akka/japi/JavaAPI.scala:
```diff
@@ -60,7 +60,7 @@ trait Predicate[T] {
  */
 @SerialVersionUID(1L)
-case class Pair[A, B](first: A, second: B) {
+case class Pair[+A, +B](first: A, second: B) {
```
I'm not sure about this... It feels like a pretty big change, that could affect a lot of things. On the other hand, I think it is right? I mean, Scala Tuples are covariant, and Pair is immutable so it should be OK...
This also brings up the question of what to do with all the other japi.TupleN types...

johanandren replied:
Why was this change introduced? This won't fly because the type is for Java, which does not have declaration-site variance.

talp-traiana replied:
javadsl.Flow.monitor wouldn't compile otherwise - Out is covariant in Flow/FlowOps but Pair isn't, so I can't return Pair<Out, FlowMonitor>

johanandren commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -1779,6 +1779,8 @@ final class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Grap
   def watchTermination[M]()(matF: function.Function2[Mat, CompletionStage[Done], M]): javadsl.Source[Out, M] =
     new Source(delegate.watchTermination()((left, right) ⇒ matF(left, right.toJava)))
 
+  def monitor: javadsl.Source[Out, Pair[Mat, FlowMonitor[Out]]] = viaMat(Flow.monitor[Out](), Keep.both[Mat, FlowMonitor[Out]])
```
"Only on Flow" means, not here either. `Source.via(Flow)` = `Source` so no need to add this.

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1934,6 +1935,8 @@ trait FlowOpsMat[+Out, +Mat] extends FlowOps[Out, Mat] {
   def watchTermination[Mat2]()(matF: (Mat, Future[Done]) ⇒ Mat2): ReprMat[Out, Mat2] =
     viaMat(GraphStages.terminationWatcher)(matF)
 
+  def monitor: ReprMat[Out, (Mat, FlowMonitor[Out])] = viaMat(Flow.monitor[Out])(Keep.both)
```
Not sure it should be here either

talp-traiana replied:
I tried putting it in FlowOps, but then I couldn't change the materialized value.
Am I missing something?

johanandren replied:
No, sorry, my bad, this _is_ the right place, I misread the folding the github view and thought it was on `RunnableGraph`. It's a flow that materializes a value - `FlowOpsMat` is right. 

Return wouldn't be a tuple but `ReprMat[Out, FlowMonitor]`, we would also want a `monitorMat` following the pattern of the other methods taking a function that allows you to choose how to handle the materialized value so that it can be called using `Keep.{left, right, both}`

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -322,6 +320,9 @@ object Flow {
     fromGraph(GraphDSL.create(sink, source)(combine) { implicit b ⇒ (in, out) ⇒ FlowShape(in.in, out.out) })
+
+  def monitor[T]: Flow[T, T, FlowMonitor[T]] =
+    fromGraph(new MonitorFlow[T])
```
I think it belongs in `FlowOps` which should make it available from the right places (and that's where all the other flow stages are available). This will make it available from instances of `Flow` and `Source` as those inherit FlowOps. 

I was wrong saying "on Flow", adding it on `FlowOps` will make it possible to `Flow[String].monitor` as well as `source.monitor`

talp-traiana commented:
@johanandren 
Sorry, I'm a little confused on where I should/shouldn't add these methods. Can you go over this list and say which of these should be implemented and which shouldn't?

- akkadsl.Flow object
- akkadsl.Flow (Flow class/instances)
- akkadsl.FlowOps (I don't think this works...)
- akkadsl.FlowOpsMat.monitor
- javadsl.Flow object (dual to the one in akkadsl.Flow)
- javadsl.Flow (class/instances, dual to akkadsl.Flow/FlowOps/FlowOpsMat)
- javadsl.Source (class/instances, dual to akkadsl.FlowOps/FlowOpsMat)

johanandren commented:
Sure, probably my fault with some unfocused reviewing.

* `FlowOpsMat.monitor` as it will put it on instances of `Source` and `Flow` in the Scala API, note that it should materialize to your new `FlowMonitor` and not a tuple. It might be useful to also provide the  `monitorMat` version, I'm not quite sure, this allows to keep a previous materialized value in the flow. One could hovever just do `source.viaMat(Flow[T].monitor)(Keep.both)` for those cases so that's why I'm not sure it is needed.
* `javadsl.Source` and `Flow` - I think you need both those as that is where they are available from the Scala-DSL, there should be automatic tests that compare the two DSLs will fail if I'm incorrect about this.

talp-traiana commented:
OK, I removed all the methods from the objects, added monitorMat (doing it "the long way" from Java would've been pretty annoying, and most of the other DSL methods have *Mat equivalents, so why not?), and cleaned up a bit. What do you think?

(And I'm still missing docs+tests, I'll probably get to those tomorrow)

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala:
```diff
@@ -269,6 +270,57 @@ object GraphStages {
     TerminationWatcher.asInstanceOf[GraphStageWithMaterializedValue[FlowShape[T, T], Future[Done]]]
 
+  private class FlowMonitorImpl[T] extends FlowMonitor[T] {
+    private val stateRef = new AtomicReference[StreamState[T]](Initialized)
```
extend AtomicReference directly

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala:
```diff
@@ -269,6 +270,57 @@ object GraphStages {
+    def setState(state: StreamState[T]): Unit = stateRef.set(state)
+
+    override def state = stateRef.get()
+  }
```
it might be possible to save on allocation of a Received for every message if you store it in an AnyRef field instead of type T, and wrapping it only back on read if needed. Roughly:
```scala
get() match {
  case s: StreamState => s
  case elem => Received(elem)
}

```

drewhk replied:
well, it trades a branch on read with an allocation on write, but I think it is worth it most of the time.

talp-traiana replied:
What happens if T == StreamState? Are we OK with that?
I mean, it would be a shame to miss out on this optimization for such an edge case, but it still feels a bit problematic.

drewhk replied:
hm, that can be parried by the onPush checking that if the grabbed elem is instance of StreamState and if it is then wrap it in a Received().

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala:
```diff
@@ -269,6 +270,57 @@ object GraphStages {
+          override def onUpstreamFinish(): Unit = {
+            super.onUpstreamFinish()
+            monitor.setState(Finished)
+          }
```
this must call finishStage otherwise there is a leak

drewhk commented:
tests are missing, but looks like a nice start!

johanandren commented:
Looks good so far. Add tests, scaladoc on the methods (look at the other ones and follow that format) and mentions in stages-overview.rst for Java and Scala and it will be done.

talp-traiana commented on akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala:
```diff
@@ -269,6 +270,59 @@ object GraphStages {
+          override def onUpstreamFinish(): Unit = {
+            super.onUpstreamFinish()
+            monitor.set(Finished)
+            completeStage()
```
@drewhk 
Now that I look at it, I'm not sure these calls are needed. Doesn't super.onFinish take care of that?

drewhk replied:
yes, they do. I had here a wrong comment which I deleted immediately after. Have you been fixing my ghost comment :)

talp-traiana replied:
Yeah, looks like it... I just saw the mail notification and did it. ^_^

talp-traiana commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMonitorSpec.scala:
```diff
@@ -0,0 +1,68 @@
+
+    "return Received after receiving a message" in {
+      val ((source, monitor), sink) =
+        TestSource.probe[Any].monitorMat()(Keep.both).toMat(Sink.ignore)(Keep.both).run()
```
@johanandren @drewhk 
I tried connecting a TestSink.probe instead of Sink.ignore so I could check that the message actually reached the sink, but it didn't work (the message was "stuck" somewhere?). Even when I removed the monitor, and connected the source to the sink directly, still nothing happened.
Any ideas why?

johanandren replied:
Could it be because you forgot to signal demand from the sink before expecting any elements coming through?

talp-traiana replied:
Doh, for some reason I thought the sink probe started with demand already seeded... Will fix it and push a new version.

johanandren commented on akka-docs/rst/java/stream/stages-overview.rst:
```diff
@@ -1067,3 +1067,15 @@ The stage otherwise passes through elements unchanged.
+^^^^^^^
+Materializes to a ``FlowMonitor`` that monitors messages flowing through or completion of the stage. The stage otherwise
+passes through elements unchanged. Note that the ``FlowMonitor`` inserts a memory barrier every time it processes an
+event, and may therefor affect performance.
```
:+1:  that you mention this!

drewhk replied:
"therefore"

johanandren commented:
LGTM

talp-traiana commented:
I think someone needs to "OK TO TEST" this?

johanandren commented:
OK TO TEST

talp-traiana commented:
Thanks. once the tests pass I'll rebase and squash (unless there are any more comments?)

drewhk commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMonitorSpec.scala:
```diff
@@ -0,0 +1,70 @@
+
+    "return Initialized for an empty stream" in {
+      val ((source, monitor), sink) =
+        TestSource.probe[Any].monitorMat()(Keep.both).toMat(TestSink.probe[Any])(Keep.both).run()
```
you can use Source.maybe and Sink.ignore as well. Doesn't really matter though, just mentioning.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala:
```diff
@@ -269,6 +270,57 @@ object GraphStages {
+          override def onPush(): Unit = {
+            val msg = grab(in)
+            push(out, msg)
+            monitor.set(if (msg.isInstanceOf[StreamState[_]]) Received(msg) else msg)
```
a test case for this is missing

talp-traiana replied:
The last test-case doesn't take care of it (FlowMonitorSpec:60)?

drewhk replied:
yes it does, my bad

drewhk commented:
LGTM

agolubev commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1933,6 +1933,22 @@ trait FlowOpsMat[+Out, +Mat] extends FlowOps[Out, Mat] {
+    * by the monitor unchanged. Note that the monitor inserts a memory barrier every time it processes an
+    * event, and may therefor affect performance.
+    */
+  def monitor: ReprMat[Out, FlowMonitor[Out]] = monitorMat()(Keep.right)
```
@drewhk, did we switch gears to allow combinators to have mat Keep.right by default? I remember we discussed this point at https://github.com/akka/akka/pull/19498#discussion_r50118558

>I see your point, but at some point we decided that the rules are:
Keep.right everywhere, except
runWith, foldWith, etc.
This makes the "matless" version not too useful though. would be interesting what others think

that's why watch watchTermination exists only with mat combinator as attribute.

drewhk replied:
No, you are right, I think this is an oversight, at least I missed this here. We should be consistent and use Keep.left.

drewhk replied:
@talp-traiana please change this to Keep.left

talp-traiana replied:
@drewhk If we ignore the materialized value of the flow, then there's no point in adding the monitor. Maybe I should just leave the 'mat' version, like watchTermination?

drewhk replied:
was thinking the same but I don't remember how we did it with watchTermination. I think it makes sense to have only the mat version.

talp-traiana replied:
As far as I can tell from the code, there is only watchTermination, and it behaves like the mat version (i.e.: takes a combiner function)

talp-traiana commented:
@drewhk So can this be merged, or is there anything else?

drewhk commented:
I need another LGTM from someone else. We are in a company meeting this week, so things are a bit slower than usual.

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMonitorSpec.scala:
```diff
@@ -0,0 +1,70 @@
+      val ((source, monitor), sink) =
+        TestSource.probe[Any].monitor()(Keep.both).toMat(TestSink.probe[Any])(Keep.both).run()
+      source.sendComplete()
+      awaitCond(monitor.state == Finished, 3.seconds)
```
I prefer `awaitAssert` since it has better error reporting if it fails

patriknw commented on akka-stream/src/main/scala/akka/stream/FlowMonitor.scala:
```diff
@@ -0,0 +1,37 @@
+  /**
+   * Stream was created, but no events have passed through it
+   */
+  case object Initialized extends StreamState[Nothing]
```
we need a java api for this one, I'm not sure what naming convention we have in streams, but some kind of method is needed in `object FlowMonitor`:

```
/**
 * JAVA API
*/
def getInitializedInstance = Initialized
```

ktoso replied:
note that this is addressed already – `initialized()` below

patriknw commented on akka-stream/src/main/scala/akka/stream/FlowMonitor.scala:
```diff
@@ -0,0 +1,37 @@
+  /**
+   * Stream completed successfully
+   */
+  case object Finished extends StreamState[Nothing]
```
java api

patriknw commented:
LGTM, aside from a few minor things

drewhk commented:
Needs a rebase and a few final fixes, then this is ready to merge!

talp-traiana commented:
@drewhk I fixed Patrik's issues in the latest commit. Is there anything else to fix (other than the rebase)?

drewhk commented:
Sorry, I missed that. I think this only needs a rebase now.

rkuhn commented:
fixes #19390

patriknw commented:
@talp-traiana needs a rebase (sorry for the inconvenience)

patriknw commented:
LGTM again

patriknw commented:
@drewhk is this ready?

drewhk commented:
I think it is yes, LGTM

ktoso commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowMonitorSpec.scala:
```diff
@@ -0,0 +1,70 @@
+package akka.stream.scaladsl
```
Sorry for nitpicking but: missing license header

ktoso commented on akka-stream/src/main/scala/akka/stream/FlowMonitor.scala:
```diff
@@ -0,0 +1,60 @@
+package akka.stream
```
here too

ktoso commented:
LGTM - nice one.
Just the missing license (we can add it). 
Shall we rebase or would you be able to @talp-traiana ?

talp-traiana commented:
@ktoso I just pushed the rebase, but hadn't noticed the comments regarding the license.

drewhk commented:
@talp-traiana see @ktoso's last comments like 
https://github.com/akka/akka/pull/19836#discussion_r58035320
https://github.com/akka/akka/pull/19836#discussion_r58035337

talp-traiana commented:
@drewhk @ktoso OK, added the license too.

ktoso commented:
Thanks a lot! We'll merge after validation

drewhk commented:
Hm:
```
[error]  * method monitor(scala.Function2)akka.stream.scaladsl.FlowOpsMat in trait akka.stream.scaladsl.FlowOpsMat is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]
```

drewhk commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1989,6 +1989,15 @@ trait FlowOpsMat[+Out, +Mat] extends FlowOps[Out, Mat] {
+    * event, and may therefor affect performance.
+    * The `combine` function is used to combine the `FlowMonitor` with this flow's materialized value.
+    */
+  def monitor[Mat2]()(combine: (Mat, FlowMonitor[Out]) => Mat2): ReprMat[Out, Mat2] =
```
this is at the wrong place. Move it to the FlowOps trait instead and then MiMa will be happy. FlowOpsMat cannot be changed due to binary compatibility.

rkuhn replied:
No, the placement is determined by whether or not a combinator changes the materialized value type. MiMa will always complain and we will always need to add exceptions for new combinators.

drewhk replied:
ok, I see. Is this safe because no-one is allowed to extend it?

patriknw replied:
and it is safe to add those exceptions because user code will never extend FlowOps, right?

drewhk replied:
We should update the comment though as it is not very helpful (it can be interpreted in various ways):
```
INTERNAL API: extending this trait is not supported under the binary compatibility rules for Akka.
```

rkuhn replied:
If this comment needs improving then please go ahead—the intention is exactly that nobody is supposed to extend it outside Akka because otherwise we would be unable to ever extend the DSL again.

drewhk replied:
I read it as you are not allowed to extend the trait as in: adding new stuff :) hence the confusion.

rkuhn replied:
Ah, I see, I was so fixated on the `extends` keyword that I didn’t consider this interpretation :-)

What about:
~~~
INTERNAL API: this trait will be changed in binary-incompatible ways for classes that are derived from it! Do not implement this interface outside the Akka code base!

Binary compatibility is only maintained for callers of this trait’s interface.
~~~

rkuhn replied:
(the same change must be made for both FlowOps and FlowOpsMat)

talp-traiana replied:
Sorry, I'm confused. Should I Roland's comment to the FlowOps and FlowOpsMat traits?
And what about the MiMa exception? I did it once, but can't remember how...

patriknw replied:
add to https://github.com/akka/akka/blob/master/project/MiMa.scala#L778
I think you'll find similar examples, and the error message from the build shows what you should add.

talp-traiana replied:
OK, let's see if this works...

talp-traiana commented:
Well, that didn't work... I tried to run mima locally, and it passed, I guess I'm doing something wrong?

patriknw commented:
Something else got merged. I have rebased this in https://github.com/akka/akka/pull/20195
